src/Compiler/Opts/ConstantFold.idr

```diff
@@ -88,6 +88,11 @@ constFold rho (CLocal fc p) = lookup fc (MkVar p) rho
 constFold rho e@(CRef fc x) = CRef fc x
 constFold rho (CLam fc x y)
   = CLam fc x $ constFold (wk (mkSizeOf [x]) rho) y
+
+-- Expressions of the type `let x := y in x` can be introduced
+-- by the compiler when inlining monadic code (for instance, `io_bind`).
+-- They can be replaced by `y`.
+constFold rho (CLet fc x inl y $ CLocal {idx = 0} _ _) = constFold rho y
```

Comment:
You should start by `constFold` the body of the let before checking
whether it is `CLocal` because that may very well reveal that it is!

Reply:
This reverses the change allowing a  `CLocal 0` being discovered for code
of the form `let z := x in let y := z in y`.

Reply:
> let z := x in let y := z in y

I added two additional functions to the test prog dealing with this kind of nested `let` statement:

```idris
%noinline
nestedLet : Nat -> Nat
nestedLet x =
  let z := x in let y := z in y

%noinline
nestedLetNoArg : Nat
nestedLetNoArg =
  let z := 10 * 10 in let y := z in y
```

Both get properly reduced:

```javascript
/* Prog.nestedLetNoArg : Nat */
const Prog_nestedLetNoArg = __lazy(function () {
 return 100n;
});

/* Prog.nestedLet : Nat -> Nat */
function Prog_nestedLet($0) {
 return $0;
}
```